tests/test_sbr_calculator.py

```diff
@@ -0,0 +1,353 @@
+        for score, expected_val in zip(elec_scores, expected_scores):
+            assert score == expected_val
+
+        sbr_val, sbr, flex_archetype = get_sbr_scores(user_inputs)
+        assert sbr_val == 7.5
+        assert sbr == "D"
+        assert flex_archetype == FlexArchetype.GOOD_FLEXER
+
+    def test_smart_meter_only(self):
+        user_inputs = UserInputs(
+            smart_meter=True,
+            smart_ev_charger=False,
+            charger_power=EVChargerPower.NONE,
+            smart_v2g_enabled=False,
+            home_battery=False,
+            battery_size=BatterySize.NONE,
+            solar_pv=False,
+            pv_inverter_size=SolarInverterSize.NONE,
+            electric_heating=False,
+            heating_source=HeatingSource.OTHER,
+            hot_water_source=HotWaterSource.OTHER,
+            secondary_heating=False,
+            secondary_hot_water=False,
+            integrated_control_sys=False,
+        )
+
+        elec_scores = calc_electrification_score(user_inputs)
+        smart_meter_score = calc_smart_meter_score(user_inputs)
+        ics_score = calc_ics_score(user_inputs)
+        elec_scores.insert(0, smart_meter_score)
+        elec_scores.insert(len(elec_scores), ics_score)
+
+        expected_scores = [1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1]
+        for score, expected_val in zip(elec_scores, expected_scores):
+            assert score == expected_val
+
+        sbr_val, sbr, flex_archetype = get_sbr_scores(user_inputs)
+        assert sbr_val == 1
+        assert sbr == "G"
+        assert flex_archetype == FlexArchetype.LOW_TECH_FLEXER
+
+    def no_smart_meter(self):
+        user_inputs = UserInputs(
+            smart_meter=False,
+            smart_ev_charger=True,
+            charger_power=EVChargerPower.CHARGER_7KW,
+            smart_v2g_enabled=True,
+            home_battery=True,
+            battery_size=BatterySize.LARGE,
+            solar_pv=True,
+            pv_inverter_size=SolarInverterSize.LT_4KW,
+            electric_heating=True,
+            heating_source=HeatingSource.HEAT_PUMP,
+            hot_water_source=HotWaterSource.HEAT_BATTERY_OR_ELEC_HOT_WATER_TANK,
+            secondary_heating=True,
+            secondary_hot_water=True,
+            integrated_control_sys=False,
+        )
+
+        elec_scores = calc_electrification_score(user_inputs)
+        smart_meter_score = calc_smart_meter_score(user_inputs)
+        ics_score = calc_ics_score(user_inputs)
+        elec_scores.insert(0, smart_meter_score)
+        elec_scores.insert(len(elec_scores), ics_score)
+
+        expected_scores = [0, 3, 3, 1, 3, -1, 4, 2, 1, 1, 1]
+        for score, expected_val in zip(elec_scores, expected_scores):
+            assert score == expected_val
+
+        sbr_val, sbr, flex_archetype = get_sbr_scores(user_inputs)
+        assert sbr_val == 1
+        assert sbr == "G"
+        assert flex_archetype == FlexArchetype.LOW_TECH_FLEXER
+
+
+def test_all_combinations():
```

Comment:
I'm not sure I like this test:

1. It's not clear what this test is trying to test for
2. If you're writing that many levels of nested for/if statements, that should be a bit of an alarm bell already!
3. Unit test is meant to test 1 small thing at a time to provide clues to where bugs are. This test is testing a lot of things at once that if broken, doesn't actually provide clues to developers where to debug, i.e. not serving its purpose.


If for some reason you really need to do that many for loop, a more pythonic way is to use itertools.product to get the cartesian product of lists:
https://note.nkmk.me/en/python-itertools-product/